### PR TITLE
Add minify support

### LINKED ISSUE #542: Built in Minify Support
#### Problem

As the title says, currently Zola doesn't have built-in html minifier support. And as for comparison, Hugo has one by invoking:

```bash
$ hugo --minify
```

It will generate production-ready minified HTML (Inline Styles + Inline Scripts) pages.

#### Idea

By borrowing the idea from Hugo we can copy like:

```bash
$ zola build --minify
```

I'm not familiar with Rust ecosystem. So maybe using https://docs.rs/html-minifier/1.1.3/html_minifier/ is possible.

### PR DESCRIPTION
**Pull request by areille**
**IMPORTANT: Please do not create a Pull Request adding a new feature without discussing it first.**

The place to discuss new features is the forum: <https://zola.discourse.group/>
If you want to add a new feature, please open a thread there first in the feature requests section.

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

Closes #542

### CODE REVIEW

File: components/config/src/config/mod.rs [hunk 1]
```diff
@@ -74,6 +74,8 @@ pub struct Config {
 
     /// Whether to compile the `sass` directory and output the css files into the static folder
     pub compile_sass: bool,
+    /// Whether to minify the html output
+    pub minify: bool,
```
Comment (Keats) [inline components/config/src/config/mod.rs]: this needs to be `minify_html`

File: components/site/src/lib.rs [hunk 2]
```diff
@@ -540,7 +541,24 @@ impl Site {
...
+                Ok(_len) => {
+                    output = std::str::from_utf8(&mut output_bytes).expect("Unable to truncate html file.").to_string();
+                },
+                Err(MinifyError{error_type, position})=> {
+                    eprintln!("Failed to truncate html at character {}:", position);
+                    eprintln!("{}", error_type.message())
+                }
+            }
+
+        }
```
Comment (Keats) [inline components/site/src/lib.rs]: you need to have the same code everywhere we are rendering a template: section and taxonomies at least

File: components/site/src/lib.rs [hunk 3]
```diff
@@ -540,7 +541,24 @@ impl Site {
...
+                minify_js: false,
+            };
+            let mut output_bytes = output.as_bytes().to_vec();
+            match truncate(&mut output_bytes, cfg) {
+                Ok(_len) => {
+                    output = std::str::from_utf8(&mut output_bytes).expect("Unable to truncate html file.").to_string();
+                },
+                Err(MinifyError{error_type, position})=> {
+                    eprintln!("Failed to truncate html at character {}:", position);
+                    eprintln!("{}", error_type.message())
```
Comment (Keats) [inline components/site/src/lib.rs]: we should bail rather printing to stderr

**Review by Keats — CHANGES_REQUESTED** [review]

**Comment by areille** [comment]
Gone to WIP status because of a problem found with `minify-html`.

It appears to turn :
`<link href="http:&#x2F;&#x2F;areille.com&#x2F;style.css" rel="stylesheet">`
into :
`<link href=http:%%areille.com%style.css rel=stylesheet>`

**Comment by areille** [comment]
@Keats this is ready for review now.

File: components/site/src/lib.rs [hunk 4]
```diff
@@ -445,6 +446,26 @@ impl Site {
         html
     }
 
+    /// Minifies html content
+    fn minify(&self, mut html: String) -> Result<String> {
+        if self.config.minify_html {
```
Comment (Keats) [inline components/site/src/lib.rs]: can you reverse the condition so it's not indented? eg `if not self.config.minify_html { return Ok(html) } ..`

File: components/site/src/lib.rs [hunk 5]
```diff
@@ -542,9 +563,14 @@ impl Site {
     pub fn render_page(&self, page: &Page) -> Result<()> {
         let output = page.render_html(&self.tera, &self.config, &self.library.read().unwrap())?;
         let content = self.inject_livereload(output);
+        let minified_content = self.minify(content).expect("Couldn't minify html file");
```
Comment (Keats) [inline components/site/src/lib.rs]: we need to remove the expect and bubble up the error

File: components/site/src/lib.rs [hunk 6]
```diff
@@ -707,7 +733,8 @@ impl Site {
         context.insert("config", &self.config);
         let output = render_template("404.html", &self.tera, context, &self.config.theme)?;
         let content = self.inject_livereload(output);
-        self.write_content(&[], "404.html", content, false)?;
+        let minified_content = self.minify(content).expect("Couldn't minify html file");
```
Comment (Keats) [inline components/site/src/lib.rs]: same here, need to remove the expect

File: components/site/src/lib.rs [hunk 7]
```diff
@@ -747,7 +774,8 @@ impl Site {
         let list_output =
             taxonomy.render_all_terms(&self.tera, &self.config, &self.library.read().unwrap())?;
         let content = self.inject_livereload(list_output);
-        self.write_content(&components, "index.html", content, false)?;
+        let minified_content = self.minify(content).expect("Couldn't minify html file");
```
Comment (Keats) [inline components/site/src/lib.rs]: same here and everywhere.

**Review by Keats — CHANGES_REQUESTED** [review]
I think we should move the minification to `write_content` so we can't forget it and no need to repeat ourselves everywhere

**Comment by areille** [comment]
I moved the `minify()` call inside `write_content()`. Hope that it is ok now !

**Comment by Keats** [comment]
Seems good, can you add a couple of tests for the minified output?

**Comment by areille** [comment]
I added a test for `index.html`. Please let me know if this is not enough and I will add some tests for other files.

EDIT : it appears to break other tests, will correct that

**Review by Keats — APPROVED** [review]

**Comment by Keats** [comment]
Thanks! I'll bump the pinned version myself for the build to pass

**Pull request merged**